cam6_4_176: COSP-RTTOV integration with CAM

jshaw35 opened the pull request:
Closes #1324 
Closes #1328 

Requires cime6.1.106 or later

I have coupled the COSP-RTTOV satellite emulator with CAM and would like to pull it into the main code repository. This update to COSP significantly extends emulation capabilities and is described here: https://egusphere.copernicus.org/preprints/2025/egusphere-2025-169/

An important note: Integrating with COSP-RTTOV with CAM requires linking an external build of RTTOV to CESM during the build step and changes to CIME. This is further complicated by the fact that the RTTOV source code is free but only available to registered users and creating a new build is slow relative to a normal CESM build, so linking an existing RTTOV build during CESM build appears to be the best solution.

CIME fork here: https://github.com/jshaw35/cime/tree/cime6.0.217_httpsbranch02_cosprttov

jshaw35 committed:
General infrastructure changes for COSP-RTTOV.

jshaw35 committed:
Kitchen sink of cospsimulator_intr changes. Everything needed for RTTOV plus swathing, memory checks, and OPAQ outputs. Needs cleaning!

jshaw35 committed:
Move assignment of allcld_ice and allcld_liq values

jshaw35 committed:
Merge tag 'cam6_4_097' into cam6_3_144_cosprttov

Merged cospsimulator_intro.F90 and bld/configure by hand. Needs testing to see if COSP-RTTOV functionality is maintained, specifically regarding the treatment of cloud water/ice and effectively water and ice droplet sizes.

jshaw35 committed:
Remove initials, change inequality checks to modern fortran, remove old dependencies, clean up obsolete cloud water fields.

jshaw35 committed:
Prepend COSP-related namelist fields with "cosp_" and add more complete descriptions. Remove commented line from bld/configure

jshaw35 committed:
update imports and Makefile following COSPv2.0 reorg

jshaw35 committed:
Correct COSP utils path in Makefile

jshaw35 committed:
Update cosp2 repo information for git fleximod

jshaw35 committed:
Address error in CESM build with COSP but not RTTOV.

jshaw35 committed:
Change cosp2 git fleximod reference to a stable hash following recommendation from cacraig

jshaw35 committed:
Missed correction

jshaw35 committed:
Update cospsimulator_intr.F90 logic to work with new namelist variables.

jshaw35 committed:
Correct allocation and assignment of cospstateIN%hgt_matrix_half.  Per cosp.F90 this array should be:
"Height of bottom interface of atm layer(km). First level contains the bottom of the top layer. Last level contains the bottom of the surface layer."
Current CAM interface includes the top of the top layer and needs to be corrected.

jshaw35 committed:
Clean up print statement

jshaw35 committed:
Merge branch 'cam_development' into cam6_3_144_cosprttov

jshaw35 committed:
At logic to masking that allows for swaths to work correctly.

jshaw35 committed:
Revert "Clean up print statement"

This reverts commit 496a7ec4b6431cfade3e5254e799c3e3c8aac022.
This was a mistake too, I meant to revert the commit that merged recent
CAM updates.

jshaw35 committed:
Revert "Merge pull request #1343 from jimmielin/hplin/vdiff_answer_change"

This reverts commit 9b98aacbfb22edac5cf46e610336e65905fb270d, reversing
changes made to 9ccd34221dd6e65e0dfae594c27bf3c9ecc9a8de.

jshaw35 committed:
Revert changes for comparison of swathed fields.

jshaw35 committed:
Clean up print statements

jshaw35 committed:
Correct cld_cal_un outfld call to outside of if statement

jshaw35 commented:
@cacraigucar I have now ported the COSP-RTTOV interface code to the cam_development branch. These changes make CAM compatible with the COSP-RTTOV branch that @dustinswales and I are currently reviewing before merging into the official COSPv2.0 repository. I've updated the CAM namelist options and logic as well.

I made the following tests for performance:
- Bit-for-bit agreement for all COSPv2.0 variables compared to the normal cam_development.
- COSPv2.0 outputs with and without satellite-like swathing are identical (e.g. swathing just puts nans in "unobserved" gridcells, answers are unchanged even though computing time should be significantly less).
- RTTOV builds correctly and output is nearly identical to the CESM2.1.5 version that I validated for the COSP-RTTOV documentation paper.

This pull request also corrects an answer-changing model error (cospsimulator_intr.F90) in the assignment of the "hgt_matrix_half" field that COSPv2.0 uses. Per cosp.F90 this array should be:

> Height of bottom interface of atm layer(km). First level contains the bottom of the top layer. Last level contains the bottom of the surface layer.

The current CAM interface sets the first level as the top of the top layer and allocates the array as length nlay+1 instead of nlay. This changes all fields that use the lidar_simulator subroutine (CALIPSO, PARASOL, etc) and RTTOV outputs.

I think that this PR is ready for a more detailed review! I will be traveling for work until August 28th and will be slow to respond during that time.

jshaw35 committed:
Update COSPv.20 tag

cacraigucar commented on .gitmodules:
```diff
@@ -98,12 +98,12 @@
+        path = src/physics/cosp2/src
+        url = https://github.com/dustinswales/COSPv2.0
         fxrequired = AlwaysRequired
         fxsparse = ../.cosp_sparse_checkout
```
This needs to brought into an official COSP repo - it can not be on a personal fork

cacraigucar commented on bld/config_files/definition.xml:
```diff
@@ -292,6 +292,14 @@ Switch to enable building COSP simulator package.  1 => build COSP.
+</entry>
+<entry id="rttov_libdir" value="">
+Directory containing RTTOV libraries. For example:
+/glade/u/home/jonahshaw/w/RTTOV2/hdfseries_build/lib/
```
Perhaps we should install the library in a TBD location on derecho.

cacraigucar commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -2470,6 +2479,169 @@ This default logical is set in cospsimulator_intr.F90
+Default: none
+</entry>
+
+<!-- Swathing input control parameters. -->
+<entry id="cosp_N_SWATHS_ISCCP" type="integer*10"  category="cosp"
```
During documentation of this feature, could we add a section on how to set up your own satellite (if it is easily done)

jshaw35 replied:
done

cacraigucar commented on src/physics/cam/diffusion_solver_cam.F90:
```diff
@@ -631,14 +631,14 @@ end function vd_lu_qdecomp
               if( itaures .eq. 1 ) then
-                 tauresx(i) = taux(i) + tauresx(i) - tauimpx(i) + tautmsx(i) + taubljx(i)
-                 tauresy(i) = tauy(i) + tauresy(i) - tauimpy(i) + tautmsy(i) + taubljy(i)
+                 tauresx(i) = taux(i) + tautmsx(i) + taubljx(i) + tauresx(i)- tauimpx(i)
+                 tauresy(i) = tauy(i) + tautmsy(i) + taubljy(i) + tauresy(i)- tauimpy(i)
               endif
 
           else
 
-             tautotx(i) = taux(i) + tautmsx(i)
-             tautoty(i) = tauy(i) + tautmsy(i)
```
I believe this is indicative of a bad merge (this and the ChangeLog).  Perhaps we can try merging to the current head of cam_development and see if they go away.  We can do that when we meet this afternoon

jshaw35 replied:
done

cacraigucar commented on src/physics/cam/cospsimulator_intr.F90:
```diff
@@ -123,6 +125,7 @@ module cospsimulator_intr
   ! ######################################################################################
   ! Default CAM namelist settings
   ! ######################################################################################
+  ! CAM
```
Comment not needed

jshaw35 replied:
done

cacraigucar commented on src/physics/cam/cospsimulator_intr.F90:
```diff
@@ -1512,6 +2011,8 @@ subroutine cospsimulator_intr_run(state, pbuf, cam_in, emis, coszrs,     &
     !! precipitation fluxes
     call pbuf_get_field(pbuf, dpflxprc_idx, dp_flxprc  )
     call pbuf_get_field(pbuf, dpflxsnw_idx, dp_flxsnw  )
+!    call pbuf_get_field(pbuf, shflxprc_idx, sh_flxprc  )
+!    call pbuf_get_field(pbuf, shflxsnw_idx, sh_flxsnw  )
```
Delete this code?

jshaw35 replied:
done

cacraigucar commented on src/physics/cam/cospsimulator_intr.F90:
```diff
@@ -1678,7 +2200,11 @@ subroutine cospsimulator_intr_run(state, pbuf, cam_in, emis, coszrs,     &
     ! ######################################################################################
     call t_startf("construct_cosp_outputs")
-    call construct_cosp_outputs(ncol, nscol_cosp, nlay, Nlvgrid, cospOUT)
+    if (allocated(rttov_configs)) then
+        call construct_cosp_outputs(ncol,nscol_cosp,nlay,Nlvgrid,rttov_Ninstruments,cospOUT,rttov_configs)
+    else
+        call construct_cosp_outputs(ncol,nscol_cosp,nlay,Nlvgrid,rttov_Ninstruments,cospOUT)
+    end if
```
Other uses of this variable are within an `if lrttov_sim` construct - should it be?

jshaw35 replied:
rttov_configs is only allocated if lrttov_sim is true, so that extra check is not needed.

cacraigucar commented on src/physics/cam/cospsimulator_intr.F90:
```diff
@@ -1569,7 +2070,27 @@ subroutine cospsimulator_intr_run(state, pbuf, cam_in, emis, coszrs,     &
     
+    ! RTTOV surface mask (consider sea ice as well)
+    ! 1: land, 0: ocean, 2: sea ice
+    rttov_sfcmask(1:ncol) = 0
+    do i=1,ncol
+       if ((cam_in%landfrac(i) > cam_in%ocnfrac(i)) .and. (cam_in%landfrac(i) > cam_in%icefrac(i))) then
+          rttov_sfcmask(i) = 1
+       else if (cam_in%ocnfrac(i) > cam_in%icefrac(i)) then
+          rttov_sfcmask(i) = 0
+       else
+          rttov_sfcmask(i) = 2
+       end if
+    end do
```
Should probably double check this entire routine, and put sections like this within an `if lrttov_sim` construct

jshaw35 replied:
done

cacraigucar commented on src/physics/cam/cospsimulator_intr.F90:
```diff
@@ -1687,69 +2213,122 @@ subroutine cospsimulator_intr_run(state, pbuf, cam_in, emis, coszrs,     &
+    cospstateIN%rttov_time(:,2)  = (ncsec - 3600 * (ncsec / 3600)) / 60 ! Remainder divided by 60 seconds per minute
+    cospstateIN%rttov_time(:,3)  = ncsec - (3600*cospstateIN%rttov_time(:,1)) - (60*cospstateIN%rttov_time(:,2)) ! Final remainder
+
+    cospstateIN%sza(1:ncol)                    = acosd(coszrs(1:ncol)) ! Hokey because we get the SZA by taking the arcosine of cos(sza), but this seems to be the variable the radiation scheme can pass.
```
Line > 133 characters - split into two or more

jshaw35 replied:
done

cacraigucar commented on src/physics/cam/cospsimulator_intr.F90:
```diff
@@ -1687,69 +2213,122 @@ subroutine cospsimulator_intr_run(state, pbuf, cam_in, emis, coszrs,     &
+    ! Multiply by 2 to go from radius to diameter, multiply 1e6 to go from meters to microns.
+    cospstateIN%DeffLiq(:,:) = 0._r8 ! Initialize for zero everywhere.
+    where ((mr_lsliq(1:ncol,1:pver) > 0._r8) .and. (mr_ccliq(1:ncol,1:pver) > 0._r8))
+        cospstateIN%DeffLiq(:,:) = 2._r8 * 1.0e6 * (mr_lsliq(1:ncol,1:pver) + mr_ccliq(1:ncol,1:pver)) / (mr_lsliq(1:ncol,1:pver) / reff_cosp(1:ncol,1:pver,I_LSCLIQ) + mr_ccliq(1:ncol,1:pver) / reff_cosp(1:ncol,1:pver,I_CVCLIQ))          
```
Line > 133 characters - break into two or more

jshaw35 replied:
done

cacraigucar commented on src/physics/cam/cospsimulator_intr.F90:
```diff
@@ -1687,69 +2213,122 @@ subroutine cospsimulator_intr_run(state, pbuf, cam_in, emis, coszrs,     &
     ! Optical inputs
     call t_startf("construct_cospIN")
-    call construct_cospIN(ncol, nscol_cosp, nlay, cospIN)
+    call construct_cospIN(ncol, nscol_cosp, nlay, rttov_Ninstruments, cospIN, emis_grey=1.0_r8) ! Apply unitary blackbody surface emissivity to be consistent with CESM physics
```
Line > 133 characters - break into two or more

jshaw35 replied:
done

cacraigucar commented on src/physics/cam/cospsimulator_intr.F90:
```diff
@@ -1978,6 +2574,40 @@ subroutine cospsimulator_intr_run(state, pbuf, cam_in, emis, coszrs,     &
+
+    ! RTTOV
+    if (lrttov_sim) then
+       do i=1,rttov_Ninstruments ! Not sure if this logical stuff is needed or not?
```
Address this comment

jshaw35 replied:
done

cacraigucar reviewed (changes requested):
Review items for us to discuss (if needed) at our meeting this afternoon
- Also, we need to add a COSP_RTTOV regression test

jshaw35 committed:
Merge tag 'cam6_4_106' into cam6_3_144_cosprttov

Merge pull request #1338 from kuanchihwang/staging/mpas-dycore-var-output

jshaw35 committed:
Revert diffusion_solver and ChangeLog

jshaw35 committed:
Change request for CAM PR

jshaw35 committed:
Merge branch 'cam_development' into cam6_3_144_cosprttov

jshaw35 committed:
Merge branch 'cam6_3_144_cosprttov' of https://github.com/jshaw35/CAM into cam6_3_144_cosprttov

jshaw35 committed:
Merge remote-tracking branch 'upstream/cam_development' into cam6_3_144_cosprttov

jshaw35 committed:
Re-add COSP2 information to .gitmodules

jshaw35 commented on src/physics/cosp2/src:
@cacraigucar I am not sure what happened here and if it needs to be corrected.

jshaw35 commented:
@cacraigucar we finally have an official COSP tag with COSP-RTTOV (2.2.0)! I have updated .gitmodules and pulled in changes from cam_development. There was a problem with merging the "src/physics/cosp2/src" repository, which now appears to be causing checks to fail. Let me know how I can help to move this forward.

cacraigucar committed:
Put back in submodule information

cacraigucar commented:
@jshaw35 - You somehow removed the submodule information - I put it back in.

jshaw35 committed:
Merge branch 'cam6_3_144_cosprttov' of https://github.com/jshaw35/CAM into cam6_3_144_cosprttov

jshaw35 committed:
Bring Makefile.in inline with COSP changes.

cacraigucar committed:
Merge tag 'cam6_4_147' into cam6_3_144_cosprttov

Made ERI test check for Gregorian+spinup compset incompatibility

jshaw35 committed:
Update cospsimulator_intr.F90 to COSP2.2.0 (part 1)

jshaw35 committed:
Add appropriate dependencies for radar simulator and shift MODIS histogram outputs to 4d for outfld call. This commit addresses a bug that occurred when cosp_lradar_sim was set to true in user_nl_cam. This bug is likely in all versions of CAM because it is unrelated to swathing and RTTOV features.

jshaw35 committed:
Merge branch 'temp_branch' into cam6_3_144_cosprttov

jshaw35 committed:
Use symbolic links for RTTOV *.a files and move dependency on $(RTTOVDIR)/build/Makefile.local directly into Makefile.rttov

jshaw35 committed:
Create COSP-RTTOV_examples for documenting usage.

cacraigucar committed:
Updates to set the RTTOV environment variable

jshaw35 committed:
Merge branch 'cam6_3_144_cosprttov' of https://www.github.com/jshaw35/CAM into cam6_3_144_cosprttov

jshaw35 commented on cime_config/buildlib:
```diff
@@ -138,14 +163,32 @@ def _build_cam(caseroot, libroot, bldroot):
             expect(os.path.isfile(libfms), "FMS library not found {}".format(libfms))
             shutil.copy(libfms, libroot)
 
+        #--------------------------------------------------------
```
@cacraigucar I think this section (lines 166 - 182) can be removed

cacraigucar replied:
Thanks - I agree.  A remnant of our attempts to get the environment variable set.

cacraigucar committed:
Remove commented out code

jshaw35 committed:
Remove debug flag from COSP_SIMULATOR masterproc call.

jshaw35 committed:
Merge branch 'cam6_3_144_cosprttov' of https://www.github.com/jshaw35/CAM into cam6_3_144_cosprttov

cacraigucar committed:
Merge tag 'cam6_4_168' into cam6_3_144_cosprttov

cam6_4_168: Mini misc tag: ZM reserved ice cleanup; zmconv_tiedtke_add typo; HEMCO_CESM external update; ccpp_scheme_utils stub update

cacraigucar committed:
Changes to get a single CAM_LINKED_LIBS to work

cacraigucar committed:
test case for RTTOV

cacraigucar committed:
Update the RTTOV library location

cacraigucar commented on src/physics/cosp2/COSP-RTTOV_examples/README.txt:
```diff
@@ -0,0 +1,74 @@
+3. If needed, update HDF5_PREFIX, FFLAGS_HDF5, LDFLAGS_HDF5, NETCDF_PREFIX, FFLAGS_NETCDF, and LDFLAGS_NETCDF to match the settings used to build RTTOV.
+
+When building CAM/CESM, make the following changes to link RTTOV.
+1. Set RTTOV as an environment variable (e.g. export RTTOV=TRUE, setenv RTTOV=TRUE)
```
@jshaw35 - Do we still need this step?

jshaw35 replied:
I think the changes you made should remove the need for this step. I assume your test no longer explicitly sets RTTOV as an env variable? I am unable to checkout the most recent hash using gitfleximod right now to confirm from my side.

nusbaume replied:
I also think this should no longer be a needed step (the python code in `buildlib` should do this automatically if `-rttov` is included in `CAM_CONFIG_OPTS`).

cacraigucar replied:
Yes, I no longer am setting an RTTOV external.  Once you are able to confirm from your side, let me know and I'll delete this line (or you can do it)

nusbaume commented on bld/config_files/definition.xml:
```diff
@@ -282,6 +282,14 @@ Switch to enable building COSP simulator package.  1 => build COSP.
 </entry>
+<entry id="rttov" valid_values="0,1" value="0">
+Switch to enable building the RTTOV radiative transfer model. With COSP.
+1 => link COSP with RTTOV with building.
```
grammar fix:


```suggestion
1 => link COSP with RTTOV when building.
```

jshaw35 replied:
Done!

nusbaume commented on bld/config_files/definition.xml:
```diff
@@ -282,6 +282,14 @@ Switch to enable building COSP simulator package.  1 => build COSP.
+</entry>
+<entry id="rttov_libdir" value="">
+Directory containing RTTOV libraries. For example:
+/glade/u/home/jonahshaw/w/RTTOV2/hdfseries_build/lib/
```
I might update this comment to use the path to the new ACL-controlled `rttov` directory as an example instead of Jonah's personal copy.

jshaw35 replied:
Done!

nusbaume commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -2463,6 +2472,169 @@ This default logical is set in cospsimulator_intr.F90
+(e.g. /glade/derecho/scratch/$USER/$CASENAME/run/). Each namelist file
+contains information specifying the simulated instrument, channels, and
+outputs. Templates and instructions can be found in the COSP-RTTOV 
+code repository.
```
Instead of pointing them to the repository, should we just point them to the local `src/physics/cosp2/COSP-RTTOV_examples` directory in the namelist description here?

jshaw35 replied:
Done!

nusbaume commented on bld/configure:
```diff
@@ -76,6 +76,8 @@ OPTIONS
                         increasing the number of advected constituents by 4.
      -cosp              Enable the COSP simulator.
+     -rttov             Enable RTTOV within the COSP simulator.
+     -rttov_libdir <str>A string containing a path to the RTTOV install.
```
Should there be a space here?


```suggestion
     -rttov_libdir <str> A string containing a path to the RTTOV install.
```

jshaw35 replied:
Done! (then this entire line was removed as rttov_libdir was a dangling path)

nusbaume commented on bld/configure:
```diff
@@ -1008,6 +1012,22 @@ if (defined $opts{'cosp'}) {
+   $cfg_ref->set('rttov', $opts{'rttov'});
+}
+elsif ( defined $opts{'rttov'} ) {
+   die "configure ERROR: rttov defined but cosp undefined. cosp_libdir also undefined \n";
```
Is this supposed to say `rttov_libdir` here?


```suggestion
   die "configure ERROR: rttov defined but cosp undefined. rttov_libdir also undefined \n";
```

jshaw35 replied:
Done! (then this entire line was removed as rttov_libdir was a dangling path)

nusbaume commented on src/physics/cosp2/Makefile.in:
```diff
@@ -137,9 +183,12 @@ cosp_cloudsat_interface.o : $(COSP_PATH)/src/src/simulator/cosp_cloudsat_interfa
 	$(F90) $(F90FLAGS) -c $<
 
-cosp_rttovSTUB.o : $(RT_PATH)/cosp_rttovSTUB.F90
+cosp_rttovSTUB.o : $(RT_PATH)/cosp_rttovSTUB.F90 
```
Why the extra space here?

jshaw35 replied:
Done!

nusbaume commented on bld/configure:
```diff
@@ -1850,6 +1870,15 @@ if ($cosp) {
 
+    # Turn on RTTOV if passed the keyword. The cosp-rttov Makefile copies rttov libraries to the $cosp_libdir path
+    if ($rttov) {
+      # Current version when copying the RTTOV libraries for each build (both HDF5 and openmp are successfully linked now):
+      $ldflags .= " -L$cosp_libdir -lrttov13_wrapper -lrttov13_mw_scatt -lrttov13_brdf_atlas -lrttov13_emis_atlas -lrttov13_other -lrttov13_parallel -lrttov13_coef_io -lrttov13_hdf -lrttov13_main ";
+      # Let the RTTOV libraries stay where they are:
+      # A more functional code would read these from Makefile.rttov in the COSP2 directory in CAM instead of hardcoding, but I don't know how to do that yet.
+      $cfg_ref->set('ldflags', $ldflags);
+      print "Adding rttov libraries as dependencies in ldflags.\n";
+    }       
```
Just adding a note here (maybe to include in a future work issue) that there are two locations (here and in `buildlib`) that are handling specialized linking for RTTOV, and so it would probably be good long-term to unify them in a single location.

nusbaume commented on src/physics/cosp2/Makefile.rttov:
```diff
@@ -0,0 +1,86 @@
+
+# Update RTTOVDIR to point to your RTTOV installation directory containing: bin  build  include  lib  mod  obj
+RTTOVDIR=/glade/u/home/jonahshaw/w/RTTOV2/hdf_build
+FFLAGS= -I$(RTTOVDIR)/mod -I$(RTTOVDIR)/include -fPIC -O3 -fopenmp
```
Just adding a note here that we should probably cleanup some of the settings here, if they are needed at all.  Maybe worth adding to the "future-work" issue?

jshaw35 replied:
Great point! I think opemmp is not needed (not compiling or using RTTOV's parallel capabilities) and likely 03 too. Should I make these changes or leave them for later since they will require retesting?

nusbaume commented on src/physics/cosp2/COSP-RTTOV_examples/compile_RTTOV.sh:
```diff
@@ -0,0 +1,97 @@
+# Create a new ifx arch file
+rm -f "RTTOV_src/build/arch/ifx"
+echo '
+FC=ifx
+FC77=ifx
+CC=gcc
```
Just noting that we should probably include in a "future-work" issue a request to generalize and cleanup this script as well (for example using Intel or GNU for both Fortran and C code, instead of one for each), or alternatively see about replacing it entirely with something like CMake.

jshaw35 replied:
The "arch" file that is created here is a direct copy of those provided in the official RTTOV source code that I have adapted to ifx. So adjusting this code makes it less consistent with other RTTOV files. I'll make a note of this in the future work section I am drafting.

nusbaume commented on src/physics/cosp2/COSP-RTTOV_examples/README.txt:
```diff
@@ -0,0 +1,74 @@
+# Instructions on how to use COSP-RTTOV in CAM
+### Jonah Shaw, 2026/02/18 (jonah.shaw@colorado.edu)
```
@jshaw35 given that you are leaving CU soon is this still an OK email for you?

jshaw35 replied:
I should have a long-term CU affiliation that starts when I leave that allows me to use my email. Is that alright?

nusbaume replied:
Yep, that's fine with us, I just wanted to double-check!  I'll go ahead and mark this comment as resolved.

nusbaume commented on src/physics/cosp2/COSP-RTTOV_examples/README.txt:
```diff
@@ -0,0 +1,74 @@
+## Installation of RTTOV
+RTTOV is a free-to-use software, but the license requires each user to register to access the source code. Users should follow official instructions to access and download RTTOV v13.2 here: https://nwp-saf.eumetsat.int/site/software/rttov/download/#Software
+
+Once you have downloaded RTTOV, follow RTTOV's instructions to compile the libraries. Building RTTOV can take ~30 minutes, so a pre-existing build is linked to CESM for each case using COSP-RTTOV. The file "compile_RTTOV.sh" can be used as a template to compile RTTOV and download example radiative transfer coefficient files. You may need to change your compile and update paths to existing NETCDF  ...
```
Typo here:


```suggestion
Once you have downloaded RTTOV, follow RTTOV's instructions to compile the libraries. Building RTTOV can take ~30 minutes, so a pre-existing build is linked to CESM for each case using COSP-RTTOV. The file "compile_RTTOV.sh" can be used as a template to compile RTTOV and download example radiative transfer coefficient files. You may need to change your compile and update paths to existing NETCDF and HDF5 libraries. NETCDF and HDF5 are used to efficiently read RTTOV coefficient files. If these libraries are not available, RTTOV can also read coefficients from .txt and binary formats, but these files are large and may take up large amounts of memory. See RTTOV v13.2 documentation for details.
```

Also is there a URL or anything to the RTTOV documentation that we could point to here?

jshaw35 replied:
Done!

nusbaume commented on src/physics/cosp2/COSP-RTTOV_examples/README.txt:
```diff
@@ -0,0 +1,74 @@
+
+## Linking RTTOV into COSP and CAM
+Once RTTOV has been built, update Makefile.rttov to point to your installation and include appropriate flags. Specifically:
+1. Set to $RTTOVDIR to point to your installation.
```
I might be a little more specific here:



```suggestion
1. Set $RTTOVDIR to point to the directory where RTTOV is installed.
```

jshaw35 replied:
Done!

nusbaume commented on src/physics/cosp2/COSP-RTTOV_examples/README.txt:
```diff
@@ -0,0 +1,74 @@
+
+cosp_lrttov_sim = .true.,
+cosp_rttov_Ninstruments = 1,
+cosp_rttov_instrument_namelists = '/glade/u/home/jonahshaw/Scripts/git_repos/COSPv2.0/driver/run/instrument_nls/cosp2_rttov_inst_CESM.txt',
```
I might update this to point to the COSP-RTTOV_examples file:
  
    
      
        Suggested change
      
    
    
      
          
            
            cosp_rttov_instrument_namelists = '/glade/u/home/jonahshaw/Scripts/git_repos/COSPv2.0/driver/run/instrument_nls/cosp2_rttov_inst_CESM.txt',
          
          
            
            cosp_rttov_instrument_namelists = '$SRCROOT/src/physics/cosp2/COSP-RTTOV_examples/cosp2_rttov_inst_CESM.txt'

jshaw35 replied:
Done!

cacraigucar committed:
Add regression test, update COSP library and have Makefile point to access control directory

cacraigucar committed:
update externals

nusbaume commented on cime_config/testdefs/testlist_cam.xml:
```diff
@@ -1138,6 +1138,15 @@
+      <machine name="derecho" compiler="intel" category="aux_cam"/>
+    </machines>
+    <options>
+      <option name="comment" >CAM6 aquaplanet w/ RTTOV</option>
```
I think this should say `CAM7`:


```suggestion
      <option name="comment" >CAM7 aquaplanet w/ RTTOV</option>
```

jshaw35 replied:
Done!

nusbaume commented on src/physics/cosp2/COSP-RTTOV_examples/README.txt:
```diff
@@ -0,0 +1,74 @@
+Each instrument requires a namelist (example "cosp2_rttov_inst_CESM.txt") to specify:
+1. The instrument itself (e.g. AIRS, IASI) and which trace gases are used in radiative transfer calculations.
+Set "rttov_srcDir" to point to your RTTOV build or the parent directory of where your RTTOV coefficients are stored.
+Set "rttov_coefDir" as the directory RTTOV coefficients are stored. "rttov_srcDir" and "rttov_coefDir" are used together when load coefficients.
```
slight grammar fix:


```suggestion
Set "rttov_coefDir" as the directory RTTOV coefficients are stored. "rttov_srcDir" and "rttov_coefDir" are used together when loading coefficients.
```

jshaw35 replied:
Done!

nusbaume commented on src/physics/cam/cospsimulator_intr.F90:
```diff
@@ -3,8 +3,8 @@ module cospsimulator_intr
-  !                       GCM inputs
-  !         Version:      v2.1.9 (October 2025)
+  !                       GCM inputs with RTTOV spectral radiation diagnostics.
+  !         Version:      v2.2.0 (December 2025)
```
Looks like it is version `V2.2.1 (May 2026)` now?

jshaw35 replied:
Done!

nusbaume reviewed (changes requested):
Thanks for bringing in this major new capability @jshaw35 and @cacraigucar!  I have some requests, but they should hopefully all be small or something we can push off as future work.  Of course if you have any questions or concerns with any of my requests just let me know.  Thanks again!

jshaw35 committed:
Merge branch 'cam6_3_144_cosprttov' of https://github.com/jshaw35/CAM into cam6_3_144_cosprttov

jshaw35 committed:
Respond to code review comments

jshaw35 committed:
Remove rttov_libdir

nusbaume reviewed (approved):
Thanks @jshaw35!  Everything looks good to me now.

Cheryl and I will work on what are hopefully some finishing touches to our official regression test tomorrow (you probably got an invite).  If everything goes well then we should hopefully have this PR officially merged in later this week.  Thanks again for all your help on it!

cacraigucar committed:
Add compilation of RTTOV

jshaw35 commented on src/physics/cosp2/Makefile.rttov:
```diff
@@ -39,11 +40,9 @@ EXT_LIB=.a
-
-LIBDIR=$(RTTOVDIR)/lib
+LIBDIR=$(RTTOVIR)/lib
 
```
This appears to be a typo

cacraigucar committed:
Fix typo in makefile

jshaw35 committed:
Merge branch 'cam6_3_144_cosprttov' of https://github.com/jshaw35/CAM into cam6_3_144_cosprttov

jshaw35 committed:
Correct changes that were reverted in 9baeaa161909a12152f270505f641f3ba28b7d25

cacraigucar committed:
Change RTTOVDIR back to path using variable

cacraigucar committed:
Merge tag 'cam6_4_173' into cam6_3_144_cosprttov

cam6_4_173: fix energy bug in physics-dynamics coupling

cacraigucar committed:
Add missing _r8

jshaw35 committed:
Merge branch 'cam6_3_144_cosprttov' of https://github.com/jshaw35/CAM into cam6_3_144_cosprttov

jshaw35 committed:
replace acosd(coszrs(1:ncol)) with acos(coszrs(1:ncol)) * 180.0_8 / acos(-1.0_8), which uses only standard Fortran functions. This converts the result from radians to degrees by multiplying by the conversion factor 180/π, where π is computed as acos(-1.0_8).

cacraigucar committed:
Change pcols to ncol for a section of code

jshaw35 committed:
Correct warnings in code compile

jshaw35 committed:
Merge branch 'cam6_3_144_cosprttov' of https://github.com/jshaw35/CAM into cam6_3_144_cosprttov

cacraigucar committed:
Add missing _r8

jshaw35 committed:
Merge branch 'cam6_3_144_cosprttov' of https://github.com/jshaw35/CAM into cam6_3_144_cosprttov

jshaw35 committed:
Add guards when assigning cospswathsIN

cacraigucar committed:
Merge tag 'cam6_4_175' into cam6_3_144_cosprttov

cam6_4_175: Complete CCPPization of CAM5 UW (diag_TKE) PBL scheme; cleanup to vertical_diffusion_tend

cacraigucar committed:
Fix run error

jshaw35 commented on src/physics/cam/cospsimulator_intr.F90-mod:
@cacraigucar did you intend to add a new file here?

cacraigucar committed:
Add dimensions to fix out of range dimensions

cacraigucar committed:
remove file which should have not been committed

jshaw35 committed:
Merge branch 'cam6_3_144_cosprttov' of https://github.com/jshaw35/CAM into cam6_3_144_cosprttov

jshaw35 committed:
Root Cause: The cfg_rttov pointer in the cosp_optical_inputs type was never initialized (nullified) in the construct_cospIN subroutine. When destroy_cospIN attempted to use ASSOCIATED() on this uninitialized pointer, the NAG compiler raised an error because the pointer's status was undefined.

Solution: Added nullify(y%cfg_rttov) in the construct_cospIN subroutine after the allocate statement. This ensures the pointer starts in a known null state, allowing ASSOCIATED() to work correctly in the destructor.

cacraigucar committed:
Add ChangeLog for cam6_4_176
